modules/gdscript/gdscript_parser.cpp

```diff
@@ -3852,11 +3855,53 @@ bool GDScriptParser::validate_annotation_arguments(AnnotationNode *p_annotation)
 		p_annotation->resolved_arguments.push_back(value);
 	}
 
+	if (p_annotation->name == SNAME("@uid")) {
+		ExpressionNode *argument = p_annotation->arguments[0];
+
+		if (argument->type != Node::LITERAL) {
+			push_error(R"(Argument 1 of annotation "@uid" must be a string literal.)", argument);
+			return false;
+		}
+
+		Variant value = static_cast<LiteralNode *>(argument)->value;
+
+		if (value.get_type() != Variant::STRING) {
+			push_error(R"(Argument 1 of annotation "@uid" must be a string literal.)", argument);
+			return false;
+		}
+
+		p_annotation->resolved_arguments.push_back(value);
+	}
+
 	// For other annotations, see `GDScriptAnalyzer::resolve_annotation()`.
 
 	return true;
 }
 
+bool GDScriptParser::uid_annotation(const AnnotationNode *p_annotation, Node *p_target, ClassNode *p_class) {
+	ERR_FAIL_COND_V(p_annotation->resolved_arguments.is_empty(), false);
+
+#ifdef DEBUG_ENABLED
+	if (this->_has_uid) {
+		push_error(R"("@uid" annotation can only be used once.)", p_annotation);
+		return false;
+	}
+#endif // DEBUG_ENABLED
+
+	const String &uid_string = p_annotation->resolved_arguments[0];
+	if (ResourceUID::get_singleton()->text_to_id(uid_string) == ResourceUID::INVALID_ID) {
+		push_error(R"(The annotated UID is invalid.)", p_annotation);
+		return false;
+	}
+
+	ClassNode *class_node = static_cast<ClassNode *>(p_target);
+	class_node->uid_string = uid_string;
+	class_node->uid_lines = Vector2i(p_annotation->start_line - 1, p_annotation->end_line - 1); // Lines start from 1, so need to subtract.
```

Comment:
Is the check required in non-debug builds?

Reply:
I guess not. No check can lead to invalid behavior, but the user is supposed to fix the bug caught in debug build.